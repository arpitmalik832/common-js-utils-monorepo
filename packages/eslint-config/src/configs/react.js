/**
 * @file React configuration.
 */
import storybookPlugin from 'eslint-plugin-storybook';

import {
  getIgnoreConfig,
  globalsConfig,
  getBaseReactConfig,
  jestConfig,
  jsdocConfig,
  mdConfig,
  cypressConfig,
} from '../baseConfigs/index.js';

/**
 * Get the React configuration.
 * @param {string[]} ignoredDirs - The directories to ignore.
 * @param {string[]} noUnresolvedIgnoreArr - The packages to ignore.
 * @returns {import('eslint').Linter.FlatConfig} The React configuration.
 * @example
 * const config = getConfig(['node_modules', 'dist']);
 */
function getConfig(ignoredDirs, noUnresolvedIgnoreArr) {
  return [
    ...getIgnoreConfig(ignoredDirs),
    ...globalsConfig,
    ...getBaseReactConfig(noUnresolvedIgnoreArr),
    ...cypressConfig,
    ...storybookPlugin.configs['flat/recommended'],
    ...jestConfig,
    ...jsdocConfig,
    ...mdConfig,
  ];
}

export default getConfig;
