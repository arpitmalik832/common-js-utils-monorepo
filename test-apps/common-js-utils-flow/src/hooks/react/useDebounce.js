// @flow

function useDebounce<D, T: VoidFunctionWithParams<D>>(
  func: T,
  timeout: number = 200,
): (...args: D[]) => void {
  let timer;
  return (...args: D[]) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func(...args);
    }, timeout);
  };
}

export default useDebounce;
