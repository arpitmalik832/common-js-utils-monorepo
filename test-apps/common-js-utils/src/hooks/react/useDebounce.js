/**
 * The useDebounce hook.
 * @file This file is saved as `useDebounce.js`.
 */

/**
 * Debounce a function to limit the rate at which it can be called.
 * @param {Function} func - The function to debounce.
 * @param {number} timeout - The time in milliseconds to wait before calling the function.
 * @returns {Function} A debounced version of the provided function.
 * @example
 * const debouncedFunction = useDebounce(myFunction, 1000);
 */
function useDebounce(func, timeout = 200) {
  let timer;
  return (...args) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func(...args);
    }, timeout);
  };
}

export default useDebounce;
