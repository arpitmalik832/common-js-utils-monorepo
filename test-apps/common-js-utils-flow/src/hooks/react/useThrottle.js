// @flow
import { useState } from 'react';

function useThrottle<D, T: VoidFunctionWithParams<D>>(
  func: T,
  limit: number = 200,
): (...args: D[]) => void {
  const [inThrottle, setInThrottle] = useState(false);

  return (...args: D[]) => {
    if (!inThrottle) {
      func(...args);
      setInThrottle(true);
      setTimeout(() => {
        setInThrottle(false);
      }, limit);
    }
  };
}

export default useThrottle;
