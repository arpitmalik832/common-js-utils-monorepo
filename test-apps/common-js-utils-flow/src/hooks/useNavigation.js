// @flow
import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router';
import {
  slices,
  beforeUnload,
  errorLog,
  log,
} from '@arpitmalik832/common-js-utils-flow';

import { BACK_CLICK, APP_UNMOUNT } from '../enums/app';

function useNavigation(): UseNavigation {
  const stack = useSelector(state => state.navigation);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleBackPress = useCallback(() => {
    if (stack.length) {
      dispatch(slices.navigationSliceActions.pop());
    } else {
      const res = navigate(-1);
      if (res instanceof Promise) {
        res
          .then(() => {
            log(BACK_CLICK.SUCCESS);
          })
          .catch(err => {
            errorLog(BACK_CLICK.ERROR, err);
          });
      }
    }
  }, [stack]);

  window.backPress = handleBackPress;

  useEffect(() => {
    beforeUnload.subscribe(() => {
      log(APP_UNMOUNT);
    });

    return () => {
      beforeUnload.unSubscribe();
    };
  }, []);

  function push(callback: VoidFunctionWithParams<mixed>) {
    dispatch(slices.navigationSliceActions.push(callback));
  }

  function pop() {
    handleBackPress();
  }

  function replace(callback: VoidFunctionWithParams<mixed>) {
    dispatch(slices.navigationSliceActions.replace(callback));
  }

  const clear = useCallback(() => {
    if (stack.length) {
      dispatch(slices.navigationSliceActions.clear());
    }
  }, [stack]);

  return { stack, push, pop, replace, clear };
}

export default useNavigation;
