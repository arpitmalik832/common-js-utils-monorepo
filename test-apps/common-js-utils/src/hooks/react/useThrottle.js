/**
 * The useThrottle hook.
 * @file This file is saved as `useThrottle.js`.
 */
import { useState } from 'react';

/**
 * Throttle a function to limit the rate at which it can be called.
 * @param {Function} func - The function to throttle.
 * @param {number} limit - The time in milliseconds to throttle calls.
 * @returns {Function} A throttled version of the provided function.
 * @example
 * const throttledFunction = useThrottle(myFunction, 1000);
 */
function useThrottle(func, limit = 200) {
  const [inThrottle, setInThrottle] = useState(false);

  return (...args) => {
    if (!inThrottle) {
      func(...args);
      setInThrottle(true);
      setTimeout(() => {
        setInThrottle(false);
      }, limit);
    }
  };
}

export default useThrottle;
