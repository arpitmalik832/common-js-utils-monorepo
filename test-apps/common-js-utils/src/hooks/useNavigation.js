/**
 * This hook handles navigation in a route.
 * @file It is saved as `useNavigation.js`.
 */
import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router';
import {
  slices,
  beforeUnload,
  errorLog,
  log,
} from '@arpitmalik832/common-js-utils-pkg';

import { BACK_CLICK, APP_UNMOUNT } from '../enums/app';

/**
 * Custom hook to handle navigation in a route.
 * @returns {object} An object containing the stack and methods to manipulate it.
 * @example
 * const { stack, push, pop, clear } = useNavigation();
 */
function useNavigation() {
  const stack = useSelector(state => state.navigation);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleBackPress = useCallback(() => {
    if (stack.length) {
      dispatch(slices.navigationSliceActions.pop());
    } else {
      const res = navigate(-1);
      if (res instanceof Promise) {
        res
          .then(() => {
            log(BACK_CLICK.SUCCESS);
          })
          .catch(err => {
            errorLog(BACK_CLICK.ERROR, err);
          });
      }
    }
  }, [stack]);

  window.backPress = handleBackPress;

  useEffect(() => {
    beforeUnload.subscribe(() => {
      log(APP_UNMOUNT);
    });

    return () => {
      beforeUnload.unSubscribe();
    };
  }, []);

  /**
   * Pushes a callback onto the stack.
   * @param {Function} callback - The callback to push onto the stack.
   * @example
   * push(() => console.log('Callback executed'));
   */
  function push(callback) {
    dispatch(slices.navigationSliceActions.push(callback));
  }

  /**
   * Pops the last item from the stack and handles back press.
   * @example
   * const { pop } = useBackPress();
   * pop();
   */
  function pop() {
    handleBackPress();
  }

  /**
   * Replaces the last item from the stack.
   * @param {Function} callback - The callback to replace the last item.
   * @example
   * replace(() => console.log('Callback executed'));
   */
  function replace(callback) {
    dispatch(slices.navigationSliceActions.replace(callback));
  }

  const clear = useCallback(() => {
    if (stack.length) {
      dispatch(slices.navigationSliceActions.clear());
    }
  }, [stack]);

  return { stack, push, pop, replace, clear };
}

export default useNavigation;
