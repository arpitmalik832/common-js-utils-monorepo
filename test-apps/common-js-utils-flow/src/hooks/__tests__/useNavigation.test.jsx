/**
 * Unit tests for useNavigation hook.
 * @file The file is saved as `useNavigation.test.jsx`.
 */
import { renderHook, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import * as reactRedux from 'react-redux';
import { slices } from '@arpitmalik832/common-js-utils-flow-pkg';

import useNavigation from '../useNavigation';

jest.mock('react-router', () => ({
  __esModule: true,
  useNavigate: () => {
    const rand = Math.random();
    if (rand < 0.3) {
      return jest.fn(() => Promise.resolve());
    }
    if (rand < 0.6) {
      return jest.fn(() => Promise.reject(new Error('an error')));
    }
    return jest.fn();
  },
}));

jest.mock('react-redux', () => ({
  __esModule: true,
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('@arpitmalik832/common-js-utils-flow-pkg', () => ({
  ...jest.requireActual('@arpitmalik832/common-js-utils-flow-pkg'),
  log: jest.fn(),
  errorLog: jest.fn(),
  beforeUnload: {
    subscribe: e => e({}),
    unSubscribe: jest.fn(),
  },
  slices: {
    navigationSliceActions: {
      clear: jest.fn(),
      pop: jest.fn(),
      push: jest.fn(),
      replace: jest.fn(),
    },
  },
}));

describe('useNavigation unit tests', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    reactRedux.useDispatch.mockReturnValue(mockDispatch);
    reactRedux.useSelector.mockImplementation(selector =>
      selector({
        navigation: [],
      }),
    );
  });

  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
    jest.clearAllMocks();
  });

  it('snapshot test', () => {
    const { result } = renderHook(() => useNavigation());

    expect(result.current).toMatchSnapshot();
  });

  it('testing push function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.push();
    expect(slices.navigationSliceActions.push).toHaveBeenCalledTimes(1);
  });

  it('testing pop function', () => {
    reactRedux.useSelector.mockImplementation(selector =>
      selector({
        navigation: [() => jest.fn()],
      }),
    );
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(1);
  });

  it('testing pop function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(0);
  });

  it('testing pop function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(0);
  });

  it('testing pop function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(0);
  });

  it('testing pop function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(0);
  });

  it('testing pop function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.pop();
    expect(slices.navigationSliceActions.pop).toHaveBeenCalledTimes(0);
  });

  it('testing clear function', () => {
    const { result } = renderHook(() => useNavigation());

    result.current.clear();
    expect(slices.navigationSliceActions.clear).toHaveBeenCalledTimes(0);
  });

  it('testing clear function', () => {
    reactRedux.useSelector.mockImplementation(selector =>
      selector({
        navigation: [() => jest.fn()],
      }),
    );
    const { result } = renderHook(() => useNavigation());

    result.current.clear();
    expect(slices.navigationSliceActions.clear).toHaveBeenCalledTimes(1);
  });

  describe('replace function', () => {
    it('should dispatch replace action with callback', () => {
      const { result } = renderHook(() => useNavigation());
      const mockCallback = jest.fn();

      result.current.replace(mockCallback);

      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(slices.navigationSliceActions.replace).toHaveBeenCalledWith(
        mockCallback,
      );
    });

    it('should work with empty stack', () => {
      reactRedux.useSelector.mockImplementation(selector =>
        selector({
          navigation: [],
        }),
      );
      const { result } = renderHook(() => useNavigation());
      const mockCallback = jest.fn();

      result.current.replace(mockCallback);

      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(slices.navigationSliceActions.replace).toHaveBeenCalledWith(
        mockCallback,
      );
    });

    it('should work with non-empty stack', () => {
      reactRedux.useSelector.mockImplementation(selector =>
        selector({
          navigation: [jest.fn()],
        }),
      );
      const { result } = renderHook(() => useNavigation());
      const mockCallback = jest.fn();

      result.current.replace(mockCallback);

      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(slices.navigationSliceActions.replace).toHaveBeenCalledWith(
        mockCallback,
      );
    });
  });
});
