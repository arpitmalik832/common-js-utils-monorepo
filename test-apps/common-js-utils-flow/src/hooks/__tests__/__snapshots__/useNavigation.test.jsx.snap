// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNavigation unit tests snapshot test 1`] = `
{
  "clear": [Function],
  "pop": [Function],
  "push": [Function],
  "replace": [Function],
  "stack": [],
}
`;
